conbench/templates/benchmark-entity.html

```diff
@@ -104,7 +104,7 @@
       {% if benchmark.stats %}
       <li class="list-group-item active">Result</li>
       <li class="list-group-item" style="overflow-y: auto;">
-        <b>timestamp</b>
+        <b>timestamp (start time)</b>
```

Comment:
I'm slightly hesitant to add this kind of strictness of "these times must be when they start" here since not all types of benchmark runs have that kind of data on a per-benchmark basis (e.g. our microbenchmarks will all have the same timestamp for either when the macro runner like `archery` is kicked off, or when the results are being adapted and uploaded, see https://conbench.ursa.dev/runs/4c9036a68cab402c94f8942634b59fca/ for example, each language has effectively one timestamp per). 

We could encourage folks to adapt these more accurately (though some of that might be an uphill battle to send PRs to google benchmark, folly, etc. in the fullness of time we should do that, but I don't think that's a today issue). But maybe leaving this ambiguous (or up to implementation specification) gives us space to be better about the precision when we can.  

Or maybe we merge the change in this PR for now and deal with any possible fallout if it happens (I think it's unlikely that anyone is relying on those columns being precise right now).

Reply:
Love the feedback, it was rather expected.

Let me summarize what we talked about so far.

In https://github.com/conbench/conbench/issues/534 I described that as a fresh Conbench user I was confused by the lack of documentation. I asked:

> is this submission time (as recorded by Conbench), or start time or end time (as recorded and freely set by the API user)?

@austin3dickey said

> I've heard @jonkeane say in the past that we're leaving "start vs end" ambiguous on purpose, to allow for flexibility between different implementations. It's another one of those tradeoffs between easy implementation and data "meaningfulness". However, that doesn't mean we can't thoughtfully decide on a way to tempt users to congregate on one definition.

As an API consumer (I have programmed with many APIs and read API documentation of various qualities) I felt sthat this is too much ambiguity. I thought that if we don't want to specify anything about the meaning of this property then we have to at least document _that_, but I didn't feel good about adding that to the spec.

I really liked Austin's

> thoughtfully decide on a way to tempt users to congregate on one definition.

because when I build code to submit data and when I _can_ set this property freely, then if the documentation tells me how to set it then this is a no-brainer to implement. In contrast, if the documentation tells me "you are free to introduce any convention here" then I get uncertain and overthink this, and try to acquire more information so that I can make a good choice.

In https://github.com/conbench/conbench/pull/623 I added this statement to the _specification_ of the `BenchmarkCreate.timestamp` property:

> A datetime string indicating the time at which the benchmark was started. 

I commented on that PR:

> This takes some mental burden away from users and I think it makes sense in most of the cases. **To advanced users it should be obvious that any time can be communicated here**. In this API spec I think it's also valuable to let users know where/how this timestamp is used (result sort oder); and so I added that.

(emphasis mine, here)

Now it's easier to reply.

> add this kind of strictness of "these times must be when they start" here since not all types of benchmark runs have that kind of data on a per-benchmark basis

Makes total sense, this general concern. But I think it's fine because we don't _actually_ introduce strictness here, but we merely describe what this property _usually_ means, removing mental burden. We did this on the specification side of things via #623 for the input, and now we're doing the corresponding part on the output. "(usually the start time)" would have maybe been more appropriate than "(start time)", but it gets a little long for little additional value.


> when the results are being adapted and uploaded, see https://conbench.ursa.dev/runs/4c9036a68cab402c94f8942634b59fca/ for example, each language has effectively one timestamp per).

Thanks for showing an example. It's good to know when and _why_ the `timestamp` property shows something vastly different from the benchmark start time, and we can see if we can or even want to do something about that in the future.

If this ever gets annoying then our discussion here is the right way towards a _better_ solution, better and _more_ specification. Maybe the answer is to not ask for a _required_ timestamp property, but to automatically keep track of the time when this was inserted in the DB (and maybe have an _optional_ `benchmark_start_time` property that has precise specification). This here is a I think an example for a metadata concept that has no well-defined use case yet.

> maybe leaving this ambiguous (or up to implementation specification) gives us space to be better about the precision when we can

In practice with people coming and going it's just hard to be precise w/o strictness in specification. Deployment-specific conventions sound a bit hard to maintain :). I want to say again that what I saw about what we have so far was confusing.

>  (I think it's unlikely that anyone is relying on those columns being precise right now)

That. I think that reliance is unlikely to happen as long as there is no rather clear expectation (via documentation, specification, and such).

Super long reply. Sorry, I know this wasn't necessary. 

I think the summary is that this corresponds to #623, and that I do think this adds a bit of hand-holding that will lead to more happy API consumers and no actual reduction in flexibility, and so I would now follow your last thought here:

> Or maybe we merge the change in this PR for now and deal with any possible fallout if it happens

Thanks for this thoughtful exchange @jonkeane :heart: